snuba/web/views.py

```diff
@@ -92,23 +94,37 @@ def check_clickhouse() -> bool:
     Checks if all the tables in all the enabled datasets exist in ClickHouse
     """
     try:
-        for name in get_enabled_dataset_names():
-            dataset = get_dataset(name)
-            for entity in dataset.get_all_entities():
-                for storage in entity.get_all_storages():
-                    clickhouse = storage.get_cluster().get_query_connection(
-                        ClickhouseClientSettings.QUERY
-                    )
-                    clickhouse_tables = clickhouse.execute("show tables")
-                    source = storage.get_schema()
-                    if isinstance(source, TableSchema):
-                        table_name = source.get_table_name()
-                        if (table_name,) not in clickhouse_tables:
-                            return False
+        datasets = [get_dataset(name) for name in get_enabled_dataset_names()]
+        entities = itertools.chain(
+            *[dataset.get_all_entities() for dataset in datasets]
+        )
+        storages = list(
+            itertools.chain(*[entity.get_all_storages() for entity in entities])
+        )
+        known_table_names = [
+            cast(TableSchema, storage.get_schema()).get_table_name()
+            for storage in storages
+            if isinstance(storage.get_schema(), TableSchema)
+        ]
+        # De-dupe clusters by host:port pairs
+        unique_clusters = {
+            storage.get_cluster().connection_tuple(): storage.get_cluster()
+            for storage in storages
+        }.values()
+
+        logger.debug(f"checking for {known_table_names} on {unique_clusters}")
+        for cluster in unique_clusters:
+            clickhouse = cluster.get_query_connection(ClickhouseClientSettings.QUERY)
+            clickhouse_tables = clickhouse.execute("show tables")
+            for table in known_table_names:
+                if (table,) not in clickhouse_tables:
+                    logger.error(f"{table} not present in cluster {cluster}")
+                    return False
```

Comment:
This does not seem right.
For each cluster you are extracting the clickhosue tables on that cluster.
Then you check that all known tables would be in `clickhouse_tables` which will fail because some tables can be only on some clusters. 
Like the metrics tables would be on the metrics cluster but not in the others, when you run show tables on the  errors cluster, the metrics tables would not be there and the check will fail.

Reply:
Sounds good, I'll construct the mapping of tables to clusters and make the check more granular.